### ISSUE TITLE
Can Versioneer be used as a dependency instead of added to the package?

### ISSUE THREAD

**Issue by felipedau** [warner/python-versioneer#145]
Hi @warner!

Thank you so much for developing Versioneer. It is one of the most useful tools I know and I use it whenever I can. There is one thing I believe that could be improved: not adding its code to the package. The problem I see with that is relying on the packages to keep it up-to-date, making them update those files as described in [Updating Versioneer](https://github.com/warner/python-versioneer#updating-versioneer).

From my understanding, `versioneer.py` has to be installed like that because its code has to exist before calling `setup()` and there is really no way to list it as a dependency. Is that the case? Because it seems that [incremental](https://github.com/hawkowl/incremental) has found a way:

> Add this to your ``setup.py``'s ``setup()`` call, removing any other versioning arguments:
> 
> ```
> 
>    setup(
>        use_incremental=True,
>        setup_requires=['incremental'],
>        install_requires=['incremental'], # along with any other install dependencies
>        ...
>    }
>  ```
> 
> Then run ``python -m incremental.update <projectname> --create`` (you will need ``click`` installed from PyPI).
> It will create a file in your package named ``_version.py`` and look like this:
> 
> ```
>    from incremental import Version
> 
>    __version__ = Version("widgetbox", 17, 1, 0)
>    __all__ = ["__version__"]
> ```
> 
> Then, so users of your project can find your version, in your root package's ``__init__.py`` add:
> 
> ```
>    from ._version import __version__
> ```

What about `_version.py`? If that specific code must really exist in that file, can't it be populated by importing from `versioneer`? (assuming that it is now installed and used just like a regular package)

Thanks!

**Comment by warner** [comment]
Yeah, I think we'd need `setup_requires=` to avoid the need for adding source files to the project tree.

I'm hesitant to use `setup_requires=` because it bypasses `pip`'s controls on where/when downloads happen, or at least it used to. When I first started Versioneer, I was looking at [peep](https://pypi.python.org/pypi/peep) as a way to force pip to only download/install/execute code that matches a hash included in a `requirements.txt` file, and at that time, if any package used `setup_requires=`, those dependencies would get installed without the hash check, violating the security goals. To maintain the ability of downstream projects to stick to this hashed-sources-only mode, I've avoided using `setup_requires=`.

But I have to admit that I haven't used that mode much myself, so maybe it isn't really that useful. And I haven't looked at pip's current hash-checking behavior: now that peep is merged into pip, maybe it's capable of controlling `setup_requires=` too? And I'd certainly agree that Versioneer's functionality would be more modular and easier to install if it were a plugin like how Incremental manages it.

**Comment by felipedau** [comment]
On Mon, Mar 06, 2017 at 11:24:52PM -0800, Brian Warner wrote:
> Yeah, I think we'd need `setup_requires=` to avoid the need for adding source files to the project tree.
> 
> I'm hesitant to use `setup_requires=` because it bypasses `pip`'s controls on where/when downloads happen, or at least it used to. When I first started Versioneer, I was looking at [peep](https://pypi.python.org/pypi/peep) as a way to force pip to only download/install/execute code that matches a hash included in a `requirements.txt` file, and at that time, if any package used `setup_requires=`, those dependencies would get installed without the hash check, violating the security goals. To maintain the ability of downstream projects to stick to this hashed-sources-only mode, I've avoided using `setup_requires=`.
> 
> And I haven't looked at pip's current hash-checking behavior: now that peep is merged into pip, maybe it's capable of controlling `setup_requires=` too? And I'd certainly agree that Versioneer's functionality would be more modular and easier to install if it were a plugin like how Incremental manages it.

peep's readme points to the [Hash-Checking Mode](https://pip.readthedocs.io/en/stable/reference/pip_install/#hash-checking-mode) which has the warning:

> Beware of the setup_requires keyword arg in setup.py. The (rare)
> packages that use it will cause those dependencies to be downloaded by
> setuptools directly, skipping pip's hash-checking.

So looks like that has not been solved yet :/

> But I have to admit that I haven't used that mode much myself, so maybe it isn't really that useful.

I agree and I think the reason is that it's the developer's responsibility to keep track of the hashes of all of the package's dependencies if they want to keep them up-to-date.

If there was a feature which you could specify signing key fingerprints, it would certainly more popular. Unfortunately it will increase the complexity as it has to handle signatures and not just hashes. I would certainly use it if there was such thing.

From the [New and Changed `setup()` Keywords](https://setuptools.readthedocs.io/en/latest/setuptools.html#new-and-changed-setup-keywords): 

> setup_requires
> 
>     A string or list of strings specifying what other distributions need
>     to be present in order for the setup script to run. setuptools will
>     attempt to obtain these (even going so far as to download them using
>     EasyInstall) before processing the rest of the setup script or
>     commands. This argument is needed if you are using distutils
>     extensions as part of your build process; for example, extensions
>     that process setup() arguments and turn them into EGG-INFO metadata
>     files.
> 
>     (Note: projects listed in setup_requires will NOT be
>     automatically installed on the system where the setup script is
>     being run. They are simply downloaded to the ./.eggs directory
>     if theyâ€™re not locally available already. If you want them to be
>     installed, as well as being available when the setup script is
>     run, you should add them to install_requires and
>     setup_requires.)

I just realized that the packages are not actually installed and that's why incremental uses both `setup_requires` and `install_requires`. With that in mind and also that after the build `_version.py` has the version in it, maybe `versioneer.py` would not have to be used at all if there was some kind of hack using `with open('package/_version.py') and `exec()`, but I'm not sure if that would work but mostly if it's too wrong to do so. (Even of we made sure that the Versioneer that generated that file is legit, it might not be acceptable)

I believe this approach would solve the problem for users installing built packages, but not from source and I do not know how to solve that.

I do not know how hard it is for pip to kind of spawn a second instance in those cases because if this would work:

```
pip install versioneer && pip install <package>
```

why isn't that be possible?

It looks like that solving the problems described above we need changes to pip and that might not be feasible.

Anyway, I also don't see people using the hash-checking mode (myself included :/ ) and maybe for them using `setup_requires` is acceptable but I'm sure there would be the ones who prefer to stay the way it is as they can verify the Versioneer they are running is really from you.

Thanks!
